docs/standard/io/snippets/pipes/pipes-from-streams/Program.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+class Program
+{
+    static async Task Main()
+    {
+        using var stream = new StreamReader("lorem-ipsum.txt");
+        using var cts = new CancellationTokenSource();
+
+        var reader = PipeReader.Create(stream.BaseStream);
+        var writer = PipeWriter.Create(
+            Console.OpenStandardOutput(), 
+            new StreamPipeWriterOptions(leaveOpen: true));
+
+        WriteUserCancellationPrompt();
+
+        var processMessagesTask = ProcessMessagesAsync(reader, writer, cts.Token);
+        var cancelProcessingTask = Task.Run(() =>
+        {
+            while (char.ToUpperInvariant(Console.ReadKey().KeyChar) != 'C')
+            {
+                WriteUserCancellationPrompt();
+            }
+
+            cts.Cancel();
+
+            reader.CancelPendingRead();
+            writer.CancelPendingFlush();
+        });
+
+        _ = await Task.WhenAny(new[] { cancelProcessingTask, processMessagesTask });
+
+        Console.WriteLine(
+            $"\n\nProcessing {(cts.IsCancellationRequested ? "cancelled" : "completed")}.\n");
+    }
+
+    static void WriteUserCancellationPrompt() =>
+        Console.WriteLine("Press 'C' to cancel processing...\n");
+
+    static async Task ProcessMessagesAsync(
```

Comment:
This should be completing the reader and writer or it may result in a memory leak.

Reply:
I assume that would happen in the `finally` after the `await reader.CompleteAsync()`?

```csharp
finally
{
    await reader.CompleteAsync();
    await writer.CompleteAsync();
}
```